### PR TITLE
VB-4084 Create Add booker, VB-4083 Create Add prisoner to booker, VB-4082 Create add visitor to a booker/prisoner

### PR DESCRIPTION
**Pull request by AledWEvans**
VB-4084 Create Add booker,
VB-4083 Create Add prisoner to booker,
VB-4082 Create add visitor to a booker/prisoner

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/integration/CreateBookerPrisonerVisitorTest.kt [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  @BeforeEach
+  fun setup() {
+    booker = createBooker(oneLoginSub = "123", emailAddress = emailAddress)
+    prisoner = createPrisoner(booker, prisonerId = "AZ1234")
+    booker.permittedPrisoners.add(prisoner)
+    bookerRepository.saveAndFlush(booker)
+  }
+
+  @Test
+  fun `when visitor dose not exist then visitor is created`() {
```
Comment (dhirajshettymoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/integration/CreateBookerPrisonerVisitorTest.kt:33]: minor typo - does. might need a find and replace.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/integration/CreateBookerTest.kt [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    responseSpec.expectStatus().isCreated
+    val dto = getBookerDto(responseSpec)
+    assertThat(dto.reference).isNotNull()
+    assertThat(dto.oneLoginSub).isNull()
+    assertThat(dto.email).isEqualTo(emailAddress)
+    assertThat(dto.permittedPrisoners).isEmpty()
+  }
+
+  @Test
+  fun `when booker dose exist then exception is thrown`() {
```
Comment (dhirajshettymoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/integration/CreateBookerTest.kt:34]: minor typo -does

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/config/HmppsPrisonVisitBookerRegistryExceptionHandler.kt [hunk 3]
```diff
@@ -126,6 +177,30 @@ class HmppsPrisonVisitBookerRegistryExceptionHandler {
...
+      .body(
+        ErrorResponse(
+          status = e.statusCode.value(),
+          userMessage = "Invalid Argument in request",
+          developerMessage = getJSRMessage(e),
+        ),
+      )
+  }
+
+  private fun getJSRMessage(e: HandlerMethodValidationException): String {
```
[the remark this answers is not in the document]
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/config/HmppsPrisonVisitBookerRegistryExceptionHandler.kt:194]: JSR = Java Specification Requests

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/service/AuthService.kt [hunk 4]
```diff
@@ -28,8 +28,8 @@ class AuthService(
       booker = bookerRepository.findByEmail(authDetail.email) ?: throw BookerNotFoundException("Booker for Email : ${authDetail.email} not found")
       booker.oneLoginSub = authDetail.oneLoginSub
       // Create reference and then save TODO this should be using post persist, we using when we stop manual entry for bookers
-      if (booker.reference.isNullOrBlank()) {
-        booker.reference = QuotableEncoder(minLength = 10).encode(booker.id)
+      if (booker.reference.isBlank()) {
```
[the remark this answers is not in the document]
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/service/AuthService.kt:31]: it can be null! it's not optional

**Review by dhirajshettymoj — APPROVED** [review]
LGTM

**Pull request merged**